clientv3/concurrency/key.go

```diff
@@ -31,34 +32,34 @@ func waitDelete(ctx context.Context, client *v3.Client, key string, rev int64) e
 	for wr = range wch {
 		for _, ev := range wr.Events {
 			if ev.Type == mvccpb.DELETE {
-				return nil
+				return &wr.Header, nil
 			}
 		}
 	}
 	if err := wr.Err(); err != nil {
-		return err
+		return nil, err
 	}
 	if err := ctx.Err(); err != nil {
-		return err
+		return nil, err
 	}
-	return fmt.Errorf("lost watcher waiting for delete")
+	return nil, fmt.Errorf("lost watcher waiting for delete")
 }
 
 // waitDeletes efficiently waits until all keys matching the prefix and no greater
 // than the create revision.
-func waitDeletes(ctx context.Context, client *v3.Client, pfx string, maxCreateRev int64) error {
+func waitDeletes(ctx context.Context, client *v3.Client, pfx string, maxCreateRev int64) (*pb.ResponseHeader, error) {
 	getOpts := append(v3.WithLastCreate(), v3.WithMaxCreateRev(maxCreateRev))
 	for {
```

Comment:
It loops because `waitDelete` will return after sensing one delete. There is no guarantee that all keys within the revision range will be deleted by that time. This code is a little old and the API is a bit stronger now so this can possibly be optimized to be single-shot on the watcher, but I don't want to change it for this patch.